source/common/upstream/upstream_impl.cc

```diff
@@ -262,34 +262,37 @@ ResourceManager& ClusterInfoImpl::resourceManager(ResourcePriority priority) con
   return *resource_managers_.managers_[enumToInt(priority)];
 }
 
-void ClusterImplBase::setInitialized() {
-  if (!initialized_ && health_checker_ && pending_initialize_health_checks_ == 0) {
+void ClusterImplBase::startInitialization() {
+  if (!initialization_started_ && health_checker_ && pending_initialize_health_checks_ == 0) {
     pending_initialize_health_checks_ = hosts().size();
-    ASSERT(pending_initialize_health_checks_ > 0);
 
     // TODO(mattklein123): Remove this callback when done.
     health_checker_->addHostCheckCompleteCb([this](HostSharedPtr, bool) -> void {
       if (pending_initialize_health_checks_ > 0 && --pending_initialize_health_checks_ == 0) {
-        setInitialized();
+        finishInitialization();
       }
     });
   }
 
-  initialized_ = true;
-  if (initialize_callback_ && pending_initialize_health_checks_ == 0) {
-    // Every time a host changes HC state we cause a full healthy host recalculation which
-    // for expensive LBs (ring, subset, etc.) can be quite time consuming. During startup, this
-    // can also block worker threads by doing this repeatedly. There is no reason to do this
-    // as we will not start taking traffic until we are initialized. By blocking HC updates
-    // while initializing we can avoid this. When HC responses for all hosts have arrived and
-    // we are about to initialize, we unblock further HC updates which has the additional effect
-    // of forcing a healthy host recalculation.
+  if (!initialization_started_ && pending_initialize_health_checks_ == 0) {
+    finishInitialization();
```

Comment:
might be worth initialization_started_ true here too, because it's weird to finishInitialization() with initialization_started_ == false.  Your call given it's a private variable - no subclasses are going to have invariants that get out of whack.